src/sgame/sg_bot_util.cpp

```diff
@@ -303,125 +450,149 @@ int BotValueOfUpgrades( gentity_t *self )
-	}
-	else if ( g_bot_lasgun.integer && BG_WeaponUnlocked( WP_LAS_GUN ) && usableCapital >= BG_Weapon( WP_LAS_GUN )->price )
+	for ( auto const& cl : classes )
 	{
-		*weapon = WP_LAS_GUN;
+		if ( cl.authorized && BotCanEvolveToClass( self, cl.item ) && BotEvolveToClass( self, cl.item ) )
+		{
+			return STATUS_SUCCESS;
+		}
 	}
-	else if ( g_bot_shotgun.integer && BG_WeaponUnlocked( WP_SHOTGUN ) && usableCapital >= BG_Weapon( WP_SHOTGUN )->price )
+
+	return STATUS_FAILURE;
+}
+
+// Allow human bots to decide what to buy
+// pre-condition:
+// * weapon is a valid pointer
+// * "upgrades" is an array of at least upgradesSize elements
+// post-conditions:
+// * "upgrades" contains a list of upgrades to use
+// * "weapon" contains weapon to use
+// * Returns number of upgrades to buy (does not includes weapons)
+//
+// will favor better armor above everything else. If it is possible
+// to buy an armor, then evaluate other team-utilies and finally choose
+// the most expensive gun possible.
+// TODO: allow bots to buy jetpack and grenades, despite the fact
+//   they can't use them (yet. For grenades it should not be that
+//   hard?): since default cVar prevent bots to buy those, that
+//   will be one less thing to change later and would have no impact.
+int BotGetDesiredBuy( gentity_t *self, weapon_t &weapon, upgrade_t upgrades[], size_t upgradesSize )
+{
+	ASSERT( self && upgrades );
+	ASSERT( self->client->pers.team == TEAM_HUMANS ); // only humans can buy
+	ASSERT( upgradesSize >= 2 ); // we access to 2 elements maximum, and don't really check boundaries (would result in a nerf)
+	int equipmentPrice = BotValueOfWeapons( self ) + BotValueOfUpgrades( self );
+	int credits = self->client->ps.persistant[PERS_CREDIT];
+	int usableCapital = credits + equipmentPrice;
+	size_t numUpgrades = 0;
+	int usedSlots = 0;
+
+	unsigned int numTeamUpgrades[UP_NUM_UPGRADES] = {};
+	unsigned int numTeamWeapons[WP_NUM_WEAPONS] = {};
+
+	ListTeamEquipment( self, numTeamUpgrades, UP_NUM_UPGRADES, numTeamWeapons, WP_NUM_WEAPONS );
+	weapon = WP_NONE;
+	for ( size_t i = 0; i < upgradesSize; ++i )
 	{
-		*weapon = WP_SHOTGUN;
+		upgrades[i] = UP_NONE;
 	}
-	else if ( g_bot_painsaw.integer && BG_WeaponUnlocked( WP_PAIN_SAW ) && usableCapital >= BG_Weapon( WP_PAIN_SAW )->price )
+
+	const equipment_t<upgrade_t> * currentArmor = nullptr;
+	for ( auto const &armor : armors )
 	{
-		*weapon = WP_PAIN_SAW;
+		if ( BG_InventoryContainsUpgrade( armor.item, self->client->ps.stats ) )
+		{
+			currentArmor = &armor;
+			break;
+		}
 	}
-	else
+
+	for ( auto const &armor : armors )
 	{
-		*weapon = WP_MACHINEGUN;
+		if ( armor.canBuyNow() && usableCapital >= armor.price()
+				&& ( usedSlots & armor.slots() ) == 0 )
+		{
+			if( !currentArmor || ( armor.price() > currentArmor->price() ) )
+			{
+				currentArmor = &armor;
+			}
+			upgrades[numUpgrades] = currentArmor->item;
+			usableCapital -= currentArmor->price();
+			usedSlots |= currentArmor->slots();
```

Comment:
There is a bug here: what if you have less used slots than before?

```suggestion
			usedSlots &= ~currentArmor->slots();
			usedSlots |= currentArmor->slots();
```

some day we'll have a proper abstraction for inventory support, but hey